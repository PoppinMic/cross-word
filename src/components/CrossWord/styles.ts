import styled from 'styled-components';
import {
  color_light_gray,
  color_white,
  color_text_button,
  color_gray_button_bg,
  color_gray_button_hover
} from '../../styles/cssVars';

export const GridWrapper = styled.div`
  padding: 2px;
  flex: 0 0 50%;
  max-width: 23em;
  margin-right: 2em;
`;

export const ButtonsWrapper = styled.div`
  order: 2;
  flex: 0 0 100%;
  margin-top: 2em;
  text-align: center;
  button {
    margin-right: 1em;
    background-color: ${color_gray_button_bg};
    color: ${color_text_button};
    border: none;
    cursor: pointer;
    display: inline-block;
    position: relative;
    outline: none;
    padding: 0.6em 1.2em;
    line-height: 1.5em;
    border-radius: 2px;
    transition: all 0.2s linear, color;
    box-shadow: none;
    user-select: none;
    &:hover {
      background-color: ${color_gray_button_hover};
      box-shadow: 0 3px 5px 2px #cbcbcb;
    }
  }
`;

export const WordListWrapper = styled.ul`
  max-width: 30%;
  padding: 0;
  li {
    list-style-position: inside;
    margin-bottom: 0.5em;
  }
`;

export const BlockWrapper = styled.span`
  position: relative;
`;

export const Block = styled.input`
  position: relative;
  /* -1px is used here to avoid double border */
  margin-left: -1px;
  margin-top: -1px;
  height: 1.5em;
  width: 1.5em;
  border: 1px solid black;
  box-sizing: border-box;
  display: inline-block;
  text-transform: uppercase;
  text-align: center;
  line-height: 1.5;
  z-index: 1;
  ${props =>
    props.disabled &&
    `
  background-color: ${color_white};
  border-color: ${color_light_gray};
  z-index:0;
  `}
`;

export const Overlay = styled.span`
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  z-index: 3;
  text-align: center;
  color: ${props => props.color};
  background-color: rgba(255, 255, 255, 0.7);
`;
