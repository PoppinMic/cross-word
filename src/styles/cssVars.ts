export const color_white = "#ffffff";
export const color_light_gray = "#dddddd";