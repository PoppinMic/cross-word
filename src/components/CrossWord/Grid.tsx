import React, { useState } from 'react';

import { mockCorrectAnswer } from '../../mockData';
import { ERROR_MSG } from '../../constants/messages';
import {
  BlockWrapper,
  Block,
  Overlay,
  GridWrapper,
  ButtonsWrapper
} from './styles';
import { color_correct_green, color_wrong_red } from '../../styles/cssVars';
import { compare2Dicts } from '../../helper/compare2Dicts';

interface IGridSource {
  gridSource: Array<string>;
}

const Grid = ({ gridSource }: IGridSource) => {
  // strings from gridSource should be same length
  // otherwise show error msg
  const gridLineLength = gridSource[0].length;
  const [answer, setAnswer] = useState({});
  const [correctAnswerDict, setCorrectAnswerDict] = useState({});

  const renderBlocks = (rowString: string, rowIndex: number) => {
    const rowResult: Array<JSX.Element | null> = [];
    if (rowString.length !== gridLineLength) {
      console.error(ERROR_MSG.WRONG_FORMAT);
      return;
    }
    for (let columnIndex = 0; columnIndex < gridLineLength; columnIndex++) {
      if (rowString[columnIndex] === '+') {
        rowResult.push(<Block key={rowIndex + columnIndex} disabled />);
      } else if (rowString[columnIndex] === '-') {
        rowResult.push(
          <BlockWrapper key={rowIndex + columnIndex}>
            <Block
              type="text"
              name={`x${rowIndex}x${columnIndex}`}
              maxLength={1}
              onChange={handleInputChange}
              value={answer[`x${rowIndex}x${columnIndex}`] || ''}
            />
            {renderValidation(`x${rowIndex}x${columnIndex}`)}
          </BlockWrapper>
        );
      } else {
        console.error(ERROR_MSG.WRONG_FORMAT);
      }
    }
    return rowResult;
  };

  const handleInputChange = e => {
    let newAnswer = { ...answer };
    newAnswer[e.currentTarget.name] = e.currentTarget.value;
    setAnswer(newAnswer);
  };

  const validateAnswers = () => {
    if (Object.keys(answer).length === Object.keys(mockCorrectAnswer).length) {
      setCorrectAnswerDict(compare2Dicts(answer, mockCorrectAnswer));
    } else {
      console.error(ERROR_MSG.INCOMPLETE);
    }
  };

  const renderValidation = dictKey => {
    if (Object.keys(correctAnswerDict).length > 0) {
      return correctAnswerDict[dictKey] ? (
        <Overlay role="img" aria-label="correct" color={color_correct_green}>
          ✔
        </Overlay>
      ) : (
        <Overlay role="img" aria-label="wrong" color={color_wrong_red}>
          ✖
        </Overlay>
      );
    }
  };

  const showCorrectAnswers = () => {
    setAnswer(mockCorrectAnswer);
  };

  const resetGrid = () => {
    setAnswer({});
    setCorrectAnswerDict({});
  };

  return (
    <React.Fragment>
      <GridWrapper>
        {gridSource.map((row, index) => (
          <div key={row + index}>{renderBlocks(row, index)}</div>
        ))}
      </GridWrapper>
      <ButtonsWrapper>
        <button onClick={validateAnswers}>Validate</button>
        <button onClick={showCorrectAnswers}>Show Correct Answers</button>
        <button onClick={resetGrid}>Try Again</button>
      </ButtonsWrapper>
    </React.Fragment>
  );
};

export default Grid;
