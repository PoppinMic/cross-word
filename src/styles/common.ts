import styled from 'styled-components';
import { color_light_gray, color_white } from './cssVars';

export const BlockWrapper = styled.span`
  position: relative;
`;

export const Block = styled.input`
  position: relative;
  margin-left: -1px;
  margin-top: -1px;
  height: 1.5em;
  width: 1.5em;
  border: 1px solid black;
  box-sizing: border-box;
  display: inline-block;
  text-transform: uppercase;
  text-align: center;
  line-height: 1.5;
  z-index: 1;
  ${props =>
    props.disabled &&
    `
  background-color: ${color_white};
  border-color: ${color_light_gray};
  z-index:0;
  `}
`;

export const Overlay = styled.span`
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  z-index: 3;
  text-align: center;
  color: ${props => props.color};
  background-color: rgba(255, 255, 255, 0.7);
`;
