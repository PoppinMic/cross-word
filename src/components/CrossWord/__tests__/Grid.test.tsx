import React from 'react';
import Enzyme, { shallow, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Grid from '../Grid';
import { Block } from '../styles';

Enzyme.configure({ adapter: new Adapter() });

const mockGridData = ['++-++', '+-+++', '---++', '-+++-', '--++-'];

let wrapper: ShallowWrapper;

beforeEach(() => {
  wrapper = shallow(<Grid gridSource={mockGridData} />);
});

describe(`<Grid /> rendering`, () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should have 25 input elements', () => {
    expect(wrapper.find(Block)).toHaveLength(25);
  });
  it('should have 10 available input elements', () => {
    expect(wrapper.find("[data-testid='availableInput']")).toHaveLength(10);
  });
  it('should have 15 disabled input elements', () => {
    expect(wrapper.find("[data-testid='disabledInput']")).toHaveLength(15);
  });
});
