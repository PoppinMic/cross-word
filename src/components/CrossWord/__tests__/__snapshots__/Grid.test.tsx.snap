// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> rendering renders correctly 1`] = `
<Fragment>
  <styled.div>
    <div
      key="++-++0"
    >
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="0"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="1"
        type="text"
      />
      <styled.span
        key="2"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x0x2"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="3"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="4"
        type="text"
      />
    </div>
    <div
      key="+-+++1"
    >
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="1"
        type="text"
      />
      <styled.span
        key="2"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x1x1"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="3"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="4"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="5"
        type="text"
      />
    </div>
    <div
      key="---++2"
    >
      <styled.span
        key="2"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x2x0"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.span
        key="3"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x2x1"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.span
        key="4"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x2x2"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="5"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="6"
        type="text"
      />
    </div>
    <div
      key="-+++-3"
    >
      <styled.span
        key="3"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x3x0"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="4"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="5"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="6"
        type="text"
      />
      <styled.span
        key="7"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x3x4"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
    </div>
    <div
      key="--++-4"
    >
      <styled.span
        key="4"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x4x0"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.span
        key="5"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x4x1"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="6"
        type="text"
      />
      <styled.input
        data-testid="disabledInput"
        disabled={true}
        key="7"
        type="text"
      />
      <styled.span
        key="8"
      >
        <styled.input
          data-testid="availableInput"
          maxLength={1}
          name="x4x4"
          onChange={[Function]}
          type="text"
          value=""
        />
      </styled.span>
    </div>
  </styled.div>
  <styled.div>
    <button
      onClick={[Function]}
    >
      Validate
    </button>
    <button
      onClick={[Function]}
    >
      Show Correct Answers
    </button>
    <button
      onClick={[Function]}
    >
      Try Again
    </button>
  </styled.div>
</Fragment>
`;
